File: other/d109. NOIP 2008 3.传纸条/main.cpp
#include <iostream>
#include <algorithm>
#define INF 100000000
using namespace std;
int n, m, grid[50][50], dp[50][50][50][50];

int rec(int xa, int ya, int xb, int yb){
    if (xa+ya+xb+yb == 0) return 0;
    if (xa < 0 || ya < 0 || xb < 0 || yb < 0)
        return -INF;
    if (xa==xb && ya==yb)
        return -INF;
    if (dp[xa][ya][xb][yb])
        return dp[xa][ya][xb][yb];
    return dp[xa][ya][xb][yb] =
        grid[xa][ya] + grid[xb][yb] + max({
            rec(xa-1, ya, xb-1, yb),
            rec(xa-1, ya, xb, yb-1),
            rec(xa, ya-1, xb-1, yb),
            rec(xa, ya-1, xb, yb-1)
        });
}

int main(){
    cin >> m >> n;
    for (int i=0; i<m; i++)
        for (int j=0; j<n; j++)
            cin >> grid[i][j];
    cout << rec(m-1, n-2, m-2, n-1);
    return 0;
}

File: other/d109. NOIP 2008 3.传纸条/main.py
m, n = map(int, input().split())
grid = [list(map(int,input().split())) for _ in range(m)]

dp = {}

def rec(x1, y1, x2, y2):
    if x1==x2==y1==y2==0:
        return 0
    if x1==x2 and y1 == y2:
        return -float('inf')
    if x1<0 or x2<0 or y1<0 or y2<0:
        return -float('inf')
    if (x1, y1, x2, y2) not in dp:
        dp[x1, y1, x2, y2] = max(
            rec(x1-1, y1, x2-1, y2),
            rec(x1-1, y1, x2, y2-1),
            rec(x1, y1-1, x2-1, y2),
            rec(x1, y1-1, x2, y2-1)
        ) + grid[x1][y1] + grid[x2][y2]
    return dp[x1, y1, x2, y2]

print(rec(m-1,n-2,m-2,n-1))